lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -3933,7 +3933,9 @@ ParseResult ConstantOp::parse(OpAsmParser &parser, OperationState &result) {
     } else if (width < value.getBitWidth()) {
       // The parser can return an unnecessarily wide result with leading
       // zeros. This isn't a problem, but truncating off bits is bad.
-      if (value.getNumSignBits() < value.getBitWidth() - width)
+      unsigned neededBits = value.isNegative() ? value.getSignificantBits()
+                                               : value.getActiveBits();
+      if (width < neededBits)
```

Comment:
Are we actually allowed to remove the leading zero if it is a positive value (which would happen here because `getActiveBits == bitWidth - numLeadingZeros`)? I think the APInt is just used as a signless bitvector and the return type actually determines how this bitvector should be interpreted. If the type is `sint` then it's two's complement. E.g., we have the binary number `01` (1 in decimal) and we truncate it to `1` then it's -1 in decimal. 1 (decimal) cannot be represented in a 1-bit signed integer. If the APInt is interpreted as unsigned, then we can truncate to `1`.

Reply:
I think it makes sense to look at the parsed value and see if that value as it was parsed can be truncated to the bit width. If it's positive it has a leading 0 bit (the binary `0b10` actually parses into APInt `010`), and if it's negative it has a leading 1 bit. That means we can always check if the user typed a negative or positive value. Then this check ensures that by cutting off leading bits we don't change the interpretation of the number, or rather, that there is no overflow.